src/rust/engine/process_execution/src/local.rs

```diff
@@ -76,10 +77,36 @@ impl super::CommandRunner for CommandRunner {
               })
                   .map(|posix_fs| Arc::new(posix_fs))
                   .and_then(|posix_fs| {
+                      let output_dirs_strings: Vec<String> = output_dir_paths
+                                                  .into_iter()
+                                                  .map(|p|
+                                                    format!(
+                                                      "{}/**",
+                                                      p
+                                                      .into_os_string()
+                                                      .into_string()
+                                                      .unwrap()
+                                                    )
+                                                  )
+                                                  .collect();
+
+                      let output_dirs_stats = posix_fs
+                          .expand(
+                            PathGlobs::create(&output_dirs_strings,
+                                              &[],
+                                              StrictGlobMatching::Ignore).unwrap()
+                          )
+                          .map_err(|e| format!("Error stating output dirs: {}", e))
+                          .wait()
```

Comment:
my intuition is that this call to `wait()` is a misuse of futures, so could use a little guidance on the right thing to do here.

Reply:
Your intuition is correct! :)

We're currently in a `Future.and_then` block, which takes a function which returns an `IntoIterator`. Handily, `Future` implements `IntoIterator`.

The general intention of an `and_then` block is that it returns either a `Result` or a `Future`, and that will just chain nicely into the following `and_then` or `map` blocks. So the structure we're looking for is:

```
future
  .and_then(|value| some_other_future)
  .and_then(|that_value| yet_another_future)
```

Accordingly, here, rather than `let output_dir_stats = posix_fs.expand(...).map_err(...).wait(...)`, if you just return the future, and you can `and_then` off of the parent, with the success value being passed as the argument to the closure:
```
.and_then(|posix_fs| {
  let output_dir_strings = ...;
  posix_fs.expand(...).map_err(...)
}).and_then(|output_dir_stats| {
  posix_fs.path_stats(...)...
})
```

It's worth noting, though, that Futures can generally run in parallel, but each `and_then` is a linearisation of code execution. If we want, we can go a level further, and allow the two expansions to happen in parallel (as they don't depend on each other), by structuring our code as something along the lines of:

```
.and_then(|posix_fs| {
  let output_dir_strings = ...;

  let output_dir_future = posix_fs.expand(...);
  let output_files_future = posix_fs.path_stats(...);
  output_dir_future.join(output_files_future) // join returns a Future which resolves either to a tuple of both success values, or the first error it encounters
}).and_then(|(output_dir_stats, output_file_stats)| {
  // ...
})
```

Reply:
thanks for the detailed notes! this was very helpful.